### PR TITLE
Bigtable: Add integration test for DirectPath fallback to and from CFE

### PR DESCRIPTION
**Pull request by igorbernstein2**
//cc @WeiranFang

This test will simulate an IPv6 outage using a faulty Netty ChannelHandler. Unfortunately this requires a reflection hack to extract the underlying NettyChannelBuilder from the ComputeEngine, which makes this test dependent on the implementation of ForwardingChannelBuilder which is marked ExperimentalApi. This isn't great, but it's better than not testing this functionality or doing some kind of man in the middle proxy or trying to implement the grpclb and dns protocols in the test.

### CODE REVIEW

**Comment by igorbernstein2** [comment]
A bit odd that coverage went down from adding a test :man_shrugging:

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Test DirectPath fallback behavior by injecting a ChannelHandler into the netty stack that will
+ * disrupt IPv6 communications.
+ */
+@RunWith(JUnit4.class)
+public class DirectPathFallbackIT {
```
Comment (elharo) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: I think the convention for integration tests in this project is that they begin IT and end Test; i.e. ITDirectPathFallbackTest

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  private AtomicInteger numBlocked = new AtomicInteger();
+  private AtomicInteger numIPv6Read = new AtomicInteger();
+
+  private ChannelFactory<NioSocketChannel> channelFactory;
+  private EventLoopGroup eventLoopGroup;
+  private BigtableDataClient instrumentedClient;
+
+  @Before
+  public void setup() throws IOException {
+    assume()
```
Comment (elharo) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: when will this not be true?

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        .withMessage("DirectPath integration tests can only run against DirectPathEnv")
+        .that(testEnvRule.env())
+        .isInstanceOf(DirectPathEnv.class);
+
+    // Create a transport channel provider that can intercept ipv6 packets.
+    channelFactory = new MyChannelFactory();
```
Comment (elharo) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: fields can be initialized directly

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    // packets.
+    assertThat(numBlocked.get()).isGreaterThan(0);
+
+    // Make sure that the client will start reading from IPv6 again by sending new requests and
+    // checking the injected IPv6 counter has been updated.
+    blackholeIPv6.set(false);
+    numIPv6Read.set(0);
+
+    // Arbitrary count that will ensure that there have been enough activity via IPv6.
+    final int MIN_COMPLETE_READ_CALLS = 40;
```
Comment (elharo) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: Local variables shouldn't be final unless required and shouldn't be named like fields

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+    while (numIPv6Read.get() < MIN_COMPLETE_READ_CALLS) {
+      for (int i = 0; i < 20; i++) {
```
Comment (elharo) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: why 20 and not 15 or 76 or any other number?

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+      delegateField.setAccessible(true);
+
+      ComputeEngineChannelBuilder gceChannelBuilder =
+          ((ComputeEngineChannelBuilder) channelBuilder);
+      Object delegateChannelBuilder = delegateField.get(gceChannelBuilder);
+
+      NettyChannelBuilder nettyChannelBuilder = (NettyChannelBuilder) delegateChannelBuilder;
+      nettyChannelBuilder.channelFactory(channelFactory);
+      nettyChannelBuilder.eventLoopGroup(eventLoopGroup);
+    } catch (NoSuchFieldException | IllegalAccessException e) {
```
Comment (elharo) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: can these exceptions simply bubble?

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/test_helpers/env/DirectPathEnv.java [hunk 7]
```diff
@@ -71,29 +77,50 @@ private DirectPathEnv(
...
               EnhancedBigtableStubSettings.defaultGrpcTransportProviderBuilder()
-                  .setPoolSize(1)
+                  .setChannelConfigurator(
+                      new ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder>() {
+                        @Override
+                        public ManagedChannelBuilder apply(ManagedChannelBuilder builder) {
+                          // Configure pick_first strategy to force a single subchannel per
+                          // ManagedChannel
+                          builder.defaultServiceConfig(
+                              ImmutableMap.of(
```
Comment (elharo) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/test_helpers/env/DirectPathEnv.java]: this is hard to follow, consider using more local variables

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 1 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: This is following the convention for bigtable:
https://github.com/googleapis/google-cloud-java/tree/master/google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 2 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: This project uses maven profiles to run integration tests against different environments. The mechanics are described in TESTING.md:
https://github.com/googleapis/google-cloud-java/blob/master/TESTING.md#testing-code-that-uses-bigtable

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 3 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: I want a new instance for every test run, so it should stay in setup

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 4 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: I promoted it to a static field

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 5 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: I added a comment explaining it a bit more.

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 6 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: No, this method gets invoked by a callback that doesn't allow checked exceptions

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/test_helpers/env/DirectPathEnv.java (hunk 7 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/test_helpers/env/DirectPathEnv.java]: Done

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 3 above)
Comment (elharo) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: This common misconception about JUnit. Every test runs with a new test object. Instance fields are not shared between test methods.

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 5 above)
Comment (elharo) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: I don't see that. Do you need to push?

Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: oops, sorry, should be there now

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java (hunk 3 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: ok, good to know. Moved to the ctor

**Comment by igorbernstein2** [comment]
All comments should be addressed, PTAL

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+        SocketAddress remoteAddress,
+        SocketAddress localAddress,
+        ChannelPromise promise)
+        throws Exception {
+
+      this.isIPv6 =
+          (remoteAddress instanceof InetSocketAddress)
+              && ((InetSocketAddress) remoteAddress).getAddress() instanceof Inet6Address;
+
+      if (!(isIPv6 && blackholeIPv6.get())) {
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: nit - I'd extract `isIPv6 && blackholeIPv6.get()` into a variable to be more descriptive

Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java]: done
